README.md
# Goalkeeper

Goalkeeper is a simple library for tracking if, and when, a Goal has
been met.  A Set of goals can be combined to check if all of them
have been completed.

This could be used as a checklist for process completion.

### Examples

Check if your application ran its daily backup

```ruby
# send the report and mark the report as sent
def backup
  # ... backup code ...
  Goalkeeper.met!("backup:#{Date.today}")
end

# elsewhere warn if any customer has not had a report sent
def did_backup_today?
  Goalkeeper.met?("backup:#{Date.today}")
end
```

Wait for a Set of Goals to be met before starting another.
```ruby
# The email can't be sent until the 'prices' and 'awards' are downloaded
def email_report
  set = Goalkeeper::Set.new.
    add("prices:#{Date.today}").
    add("awards:#{Date.today}")

  if set.met?
    log.info "Downloads completed by #{set.met_at}"
    deliver
  else
    log.info "#{set.unmet.map(&:label).join(', ')} have not been downloaded"
    retry_later
  end
end

# elsewhere the code to get the data and mark it complete
def download_prices
  # ... download code ...
  Goalkeeper.met!('prices:#{Date.today}')
end

def download_awards
  # ... download code ...
  Goalkeeper.met!('awards:#{Date.today}')
end
```

### Installation

Add this line to your application's Gemfile:

```ruby
gem 'goalkeeper'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install goalkeeper

### Usage

Its goes likes this...

You might never need to reference the Goal class.  Most everything can be done via Goalkeeper and Goalkeeper::Set.

```ruby
Goalkeeper.met? "label" #=> false
Goalkeeper.met! "label" #=> <Goalkeeper::Goal>
Goalkeeper.met? "label" #=> true
```

But lets have a granular example.  Lets ensure we wakeup New Years Day 2020.
The Goal will be named 'wakeup:2020-01-01'

```ruby
g = Goalkeeper::Goal.new('wakeup:2020-01-01')
g.met? #=> false
```

Time flies... it is New Years Day 2020.

```ruby
g.met! # or Goalkeeper.met!('wakeup:2020-01-01')
g.met? #=> true
g.met_at #=> 2020-01-01 05:01:31 -0500
```

Now if our application checks our goal, it will be met.

```ruby
Goalkeeper.met?('wakeup:2020-01-01') #=> true
# or
Goalkeeper::Goal.new('wakeup:2020-01-01').met? #=> true
```

Note: Once a Goal is 'met' the 'met\_at' timestamp will not change, unless
'clear!' is called.

We are probably only interested in this goal being complete for a limited
time, so it will expire and be removed from Redis.

```ruby
g.ttl #=> 86400 (1 day)
```

If you need to reference the Redis key

```ruby
g.key #=> Goalkeeper:wakeup:2020-01-01
```

Finally deleting the Goal is simple

```ruby
g.clear!
g.met? #=> false
```

#### Sets

Perhaps you have a series of Goals you want to track, and see if they all have been met, or not.

```ruby
set = Goalkeeper::Set.new
set.add('goal1').add('goal2')
set.met? #=> false
```

Lets have 1 goal met:

```ruby
Goalkeeper.met!('goal1')
```

But our set is not met yet

```ruby
set.met? #=> false
```

See which goals are met, or unmet

```ruby
set.met #=> [#<Goalkeeper::Goal @label="goal1">]
set.unmet #=> [#<Goalkeeper::Goal @label="goal2">]
```

Lets complete our set.

```ruby
Goalkeeper.met!('goal1')
set.met? #=> true
```

See the time the final goal was met

```ruby
set.met_at #=> 2015-01-01 08:02:15 -0500
```

#### Customization

Customize the redis client by setting it in your application

```ruby
Goalkeeper.redis = your_redis_client
```

Each record has a default expiration of 24 hours, but this can be modified.

```ruby
Goalkeeper.expiration = number_of_seconds
```

Redis keys are stored under the default namespace of "Goalkeeper:". The
namespace can be configured:

```ruby
Goalkeeper.namespace = string
```

## API

#### Goal
| method     | description                                                | example                                  |
| --------   | ------                                                     | -----                                    |
| new(label) | Creates a new Goal with the unique label                   | `g = Goal.new('process')`                |
| met?       | Returns true if the Goal has been met                      | `g.met? #=> false`                       |
| met!       | Marks the goal as completed with a timestamp               | `g.met! #=> Time.now`                    |
| met\_at    | Nil if unmet, otherwise returns the time the Goal was met. | `g.met_at #=> 2015-02-09 22:24:07 -0500` |
| ttl        | The Redis ttl value if there is a record for the goal.     | `g.ttl #=> 86400`                       |
| key        | The key used to store the record in Redis.                 | `g.key #=> "Goalkeeper:process"`         |
| clear!     | Deletes the `met` record.  Goal is now unmet.              | `g.clear!`                               |


#### Set
| method     | description                                                             | example                                  |
| --------   | ------                                                                  | -----                                    |
| new        | creates an empty set of goals.                                          | `s = set.new`                            |
| add(label) | Adds a new Goal the Set. _chainable_                                    | `s.add('process.a').add('process.b')`    |
| met?       | Returns true if _all_ the Set's Goals have been met.                    | `s.met? #=> false`                       |
| met\_at    | Returns the most recent met_at for the Set's Goals. Nil if no Goal met. | `s.met_at #=> 2015-02-09 22:24:07 -0500` |
| met        | Returns a new Set with all Goals which have been met.                   | `s.met #=> Set(...)`                     |
| unmet      | Returns a new Set with all Goals which have not been met.               | `s.unmet` #=> Set(...)                   |

### Goalkeeper
| method        | description                          | example |
| --------      | ------                               | -----   |
| ::met!(label) | Creates a new Goal and marks it met. | `Goalkeeper.met!('process') #=> <Goalkeepr::Goal>`
| ::met?(label) | Checks if the label has been met or not | `Goalkeeper.met?('process') #=> true`

## Contributing

1. Fork it ( https://github.com/jweir/goalkeeper/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGES.md
# 0.3.3

Support higher version of the Redis gem

# 0.3.2

* Enhancement
  * `Goal.new` auto joins multiple arguments into a single label


test/test_helper.rb
require 'minitest/autorun'
require 'minitest/pride'
require './lib/goalkeeper'
require './test/support/redis_instance'

# Start up Redis on a random port
Goalkeeper.redis = RedisInstance.run!


test/goalkeeper_test.rb
require 'test_helper'

describe Goalkeeper do
  before do
    Goalkeeper.redis.flushdb
  end

  describe "::redis" do
    it "returns the Redis client" do
      assert Goalkeeper.redis.is_a?(Redis)
    end
  end

  describe "::namespace" do
    it "defaults to Goalkeeper" do
      assert_equal "Goalkeeper", Goalkeeper.namespace
    end

    it "can be user defined" do
      ns = Goalkeeper.namespace

      Goalkeeper.namespace = "NewNamespace"
      assert_equal "NewNamespace", Goalkeeper.namespace
      goal = Goalkeeper::Goal.new("x")
      assert_equal "NewNamespace:x", goal.key

      # reset
      Goalkeeper.namespace = ns
    end
  end

  let(:goal) {  Goalkeeper::Goal.new('a:b') }

  describe "::met!" do
    it "creates a goal and flags it met" do
      Goalkeeper.met!('a','b')
      assert goal.met?
    end
  end

  describe "::met?" do
    it "creates a goal and checks if it is met" do
      assert ! Goalkeeper.met?('a','b')
      goal.met!
      assert Goalkeeper.met?('a','b')
    end
  end
end


test/goalkeeper/set_test.rb
require 'test_helper'

describe Goalkeeper::Set do
  before do
    Goalkeeper.redis.flushdb
  end

  let(:goals) { Goalkeeper::Set.new }

  describe "#add" do
    it "creates a Goal from the label" do
      goals.add("a:1")
      assert_equal 1, goals.size
      assert_equal "a:1", goals[0].label
    end

    it "accepts an option expiration" do
      goals.add("a:1", expiration: 20)
      assert_equal 20, goals.first.expiration
    end

    it "should return itself (so it is chainable)" do
      assert_equal goals, goals.add("a:1")
    end
  end

  it "behaves as a unique set" do
    goals << Goalkeeper::Goal.new("a")
    goals << Goalkeeper::Goal.new("b")
    goals << Goalkeeper::Goal.new("a")
    goals.push Goalkeeper::Goal.new("a")

    assert_equal 2, goals.size
  end

  it "ignores insertion of nonGoals" do
    goals << "a"
    goals.push 1, 2, :a
    assert_equal 0, goals.size
  end

  describe "with goals" do
    before do
      goals.add("x").add("y")
    end

    describe "#met" do
      it "returns all Goals which have been met" do
        assert goals.met.empty?
        goals[0].met!
        assert_equal ["x"], goals.met.map(&:label)
        goals[1].met!
        assert_equal(%w( x y ), goals.met.map(&:label))

        assert goals.met.is_a?(Goalkeeper::Set)
      end
    end

    describe "#unmet" do
      it "returns all Goals which have not been met" do
        assert_equal(%w( x y ), goals.unmet.map(&:label))
        goals[0].met!
        assert_equal ["y"], goals.unmet.map(&:label)
        goals[1].met!
        assert goals.unmet.empty?

        assert goals.unmet.is_a?(Goalkeeper::Set)
      end
    end

    describe "#met?" do
      it "is true when all Goals have been met" do
        assert !goals.met?
        goals.each(&:met!)
        assert goals.met?
      end
    end

    describe "#met_at" do
      before do
        goals << Goalkeeper::Goal.new("a")
        goals << Goalkeeper::Goal.new("b")
        goals << Goalkeeper::Goal.new("a")
      end

      it "is nil unless all Goals are met" do
        goals[0].met!
        goals[1].met!
        assert_nil goals.met_at
      end

      it "is the most recent met_at from the Goals" do
        t = Time.now
        Time.stub(:now, t + 1000) do
          goals.first.met!
        end
        # meet the reset
        goals.each(&:met!)

        assert_equal((t + 1000).to_a, goals.met_at.to_a)
      end
    end
  end
end


lib/goalkeeper/set.rb
module Goalkeeper
  # Set is an Array of Goals to simplify tracking multiple goals.
  # Since Set is an array, you have all Array methods available.
  #
  # Create a new set
  #   myset = Goalkeeper::Set.new
  #
  # Add Goals you want to check for completion
  #   myset.add("job1").add("job2")
  #   myset.size
  #   #=> 2
  #
  # Check if all the goals are completed
  #   myset.met?
  #   #=> false
  #
  # Get the unmet Goals
  #   myset.unmet
  #   #=> [...]
  #
  # Get the met Goals
  #   myset.met
  #   #=> [...]
  #
  # Iterate all Goals
  #   myset.each {|goal| ...}
  #   myset.map  {|goal| ...}
  class Set < Array
    def initialize
      super
    end

    # Creates a new Goal.
    # see Goal#initialize for usage
    def add(*args)
      push(Goal.new(*args))
      self
    end

    # met? returns true if all Goals in the set have been met.
    def met?
      unmet.empty?
    end

    # unmet returns a Set with the all the Goals that have not been met.
    def unmet
      subset(select { |g| !g.met? })
    end

    # met returns a Set with the all the Goals that have been met.
    def met
      subset(select { |g| g.met? })
    end

    # nil if this set is not met?
    # otherwise returns the met_at Time for the last met goal
    def met_at
      if met?
        self.map(&:met_at).sort.last
      else
        nil
      end
    end

    def <<(other)
      if other.is_a?(Goal) && !include?(other)
        super 
      else
        false
      end
    end

    def push(*others)
      others.each do |o|
        self << o
      end
    end

    protected

    def subset(set)
      Goalkeeper::Set.new.tap do |s|
        set.each {|i| s << i}
      end
    end
  end
end


lib/goalkeeper/goal.rb
module Goalkeeper
  # Goal represents a label which has either been +met+ or not.
  # 
  class Goal
    # The unique label to identify this Goal. There is no logic to check that the
    # label is unique.
    attr_reader :label

    # the TTL value for the Redis record.  Defalts to Goalkeeper.expiration
    attr_reader :expiration

    # +label+ is a unique string to identify this Goal. If multiple args are passed they are joined and seperated by ':'
    # +expiration+ number secconds. This can be set to override the gobal expiration.
    def initialize(*label, expiration: Goalkeeper.expiration)
      @label = label.join(":")
      @expiration = expiration
    end

    def met!
      write unless met?
      self
    end

    def met?
      !read.nil?
    end

    # Time the goal was completed.
    # WARNING retuns nil if the job is not met
    def met_at
      return Time.parse(read) if met?
      nil
    end

    # clear! removes the met state of the Goal.
    def clear!
      Goalkeeper.redis.del(key)
    end

    # a namespaced key for the goal
    def key
      "#{Goalkeeper.namespace}:#{label}"
    end

    # ttl returns the time to live on the redis key
    def ttl
      Goalkeeper.redis.ttl(key)
    end

    # All Goalkeeper::Goals with the same label are equal
    def ==(other)
      other.is_a?(Goalkeeper::Goal) && other.label == label
    end

    protected

    def write
      Goalkeeper.redis.set(key, Time.now)
      Goalkeeper.redis.expire(key, expiration)
    end

    def read
      Goalkeeper.redis.get(key)
    end
  end
end


lib/goalkeeper.rb
require 'forwardable'
require 'redis'
require 'time' # for Time.parse

# Goalkeeper provides methods to track if specific events(Goals) have been
# completed(met).
#
# Its goes likes this..
#    
# Lets ensure we wakeup New Years Day 2020.  The goal will be called 'wakeup:2020-01-01'
#   g = Goalkeeper::Goal.new('wakeup:2020-01-01')
#   g.met? #=> false
#
# Time flies... it is New Years Day 2020.
#   g.met! # or Goalkeeper.met!('wakeup:2020-01-01')
#   g.met? #=> true
#   g.met_at #=> 2020-01-01 05:01:31 -0500
#
# Now if our application checks our goal, it will be met.
#   Goalkeeper::Goal.new('wakeup:2020-01-01').met? #=> true
#   Goalkeeper.met?('wakeup:2020-01-01') #=> true
#   
# Note: Once a Goal is 'met' the 'met_at' timestamp will not change, unless
# 'clear!' is called.
#
# We are probably only interested in this goal being complete for a limited
# time, so it will expire and be removed from Redis.
#   g.ttl #=> 86400 (1 day)
#
# If you need to reference the Redis key
#   g.key #=> Goalkeeper:wakeup:2020-01-01
#
# Finally clear the Goal is simple
#   g.clear!
#
# === Sets
#
# Perhaps you have a series of Goals you want to track, and see if they all have been met, or not.
#
#   set = Goalkeeper::Set.new
#   set.add('goal1').add('goal2')
#   set.met? #=> false
#
# Lets have 1 goal met:
#   Goalkeeper.met!('goal1')
#
# But our set is not met yet
#   set.met? #=> false
#
# See which goals are met, or unmet
#   set.met #=> [#<Goalkeeper::Goal @label="goal1">]
#   set.unmet #=> [#<Goalkeeper::Goal @label="goal2">]
#
# Lets complete our set.
#   Goalkeeper.met!('goal1')
#   set.met? #=> true
#
# See the time the final goal was met
#   set.met_at #=> 2015-01-01 08:02:15 -0500
#
# === Customization
#
# Customize the redis client by setting it in your application
#   Goalkeeper.redis = your_redis_client
#
# Each record has a default expiration of 24 hours, but this can be modified.
#   Goalkeeper.expiration = number_of_seconds
#
# Redis keys are stored under the default namespace of "Goalkeeper:". The
# namespace can be configured:
#   Goalkeeper.namespace = string
#
module Goalkeeper
  require "goalkeeper/version"
  require "goalkeeper/goal"
  require "goalkeeper/set"

  # Set the Redis client to a non default setting
  def self.redis=(redis)
    @redis = redis
  end

  def self.redis
    @redis ||= Redis.new
  end

  # Creates a persistent Goal market with the given label.
  def self.met!(*label)
    Goal.new(*label).met!
  end

  def self.met?(*label)
    Goal.new(*label).met?
  end

  # The TTL set for each met Goal record created in Redis
  # Default is 24 hours
  def self.expiration
    @expiration ||= 24 * 60 * 60
  end

  # Overwrite the default expiration
  def self.expiration=(number_of_seconds)
    @expiration = number_of_seconds
  end

  def self.namespace
    @namespace ||= "Goalkeeper"
  end

  def self.namespace=(ns)
    @namespace = ns
  end
end
